modules/dns_dnscmd/dns_dnscmd_main.rb

```diff
@@ -85,5 +86,45 @@ def report msg, response, error_only
       logger.error "Dnscmd failed:\n" + (response.is_a?(Array) ? response.join("\n") : "Response was not an array! #{response}")
       raise Proxy::Dns::Error.new("Unknown error while processing '#{msg}'")
     end
+
+    def match_zone(fqdn)
+      dns_zones = ::Proxy::Dns::Plugin.settings.dns_zones || enum_dns_zones
+      found_match = false
+      weight = 0 # sub zones might be independent from similar named parent zones; use weight for longest suffix match
+      matched_zone = fqdn.sub(/[^.]+./,'') #Fixme: breaks PTR - Fallback
+
+      if dns_zones.is_a?(Array)
+        dns_zones.each do |zone|
+          zone_labels = zone.split(".").reverse
+          zone_weight = zone_labels.length
+          fqdn_labels = fqdn.split(".")
+          fqdn_labels.shift
+          is_match = zone_labels.all? { |zone_label| zone_label == fqdn_labels.pop }
+          # match only the longest zone suffix
+          if is_match && zone_weight >= weight
+            matched_zone = zone
+            weight = zone_weight
+            found_match = true
+          end
+        end
+      end
+      unless found_match
+        logger.warn "Unable to determine dns zone for dnscmd.exe, falling back to #{matched_zone}"
+      else
+        logger.debug "Matched dns zone for dnscmd.exe, #{matched_zone}"
+      end
+      matched_zone
+    end
+
+    def enum_dns_zones
```

Comment:
I patched my production env with this function; does its job. Maybe we *should* run it on every match zone call as it would be the most transparent for the user.
I just timed it with 30 Zones. We can also almost half the execution time by not calling `cmd.exe`:

```powershell
PS C:\Users\Administrator> Measure-Command {c:\windows\system32\cmd.exe /c c:\Windows\System32\dnscmd.exe localhost /EnumZones}

TotalMilliseconds : 42,897


PS C:\Users\Administrator> Measure-Command {c:\Windows\System32\dnscmd.exe localhost /EnumZones}

TotalMilliseconds : 27,3026
```